pkg/location/mutate/mutations.go

```diff
@@ -0,0 +1,29 @@
+package mutate
+
+import (
+	"fmt"
+
+	"github.com/GoogleContainerTools/kpt/pkg/location"
+)
+
+type IdentifierSetter interface {
+	SetIdentifier(identifier string) (location.Reference, error)
+}
+
+func Identifier(ref location.Reference, identifier string) (location.Reference, error) {
+	if ref, ok := ref.(IdentifierSetter); ok {
+		return ref.SetIdentifier(identifier)
+	}
+	return nil, fmt.Errorf("changing identifier not supported for reference: %v", ref)
+}
+
+type LockSetter interface {
```

Comment:
Great question! Looking ahead a bit (and based on how it would be used for cli-oci feature) a Reference and ReferenceLock would be used in a few places: 

There's the Reference that's a [DIR] argument or "-" stdio stream. That's location info for direct read/write of package content.

Once package is read in, there's the Upstream Reference/ReferenceLock representation of the upstream/upstreamLock. Those are always driven by the kptfile contents and will be the remote git/oci type.

Those are the simple cases. The more interesting one like you mentioned is like what happens in the oci cli commands that were added "kpt pkg pull/push". There an Origin ReferenceLock added to the kptfile as it's pulled (it's purely client-side data and is removed again when it's pushed).

So in that case the Reference of Dir{} or stream will get you package content, and the Upstream will be where it was cloned from originally, and the Origin ReferenceLock will be where it was pulled from. (Similar to the origin of git folder, really.)

Last case worth mentioning --- if you have a location Reference that's directly to git/oci (and not to a pulled dir copy) that's where you'd expect that remote Reference to source the content directly. In that case Upstream would come from the kptfile of course and be just that data, but if you ask for the Origin of _that_ result there you'll expect the ReferenceLock to be the commit/digest on top of the remote Reference that was pulled.

TBH, the idea of Origin was Brian's idea earlier. He mentioned that if pkg pull added origin data, it would give a default push location. And also that saving the digest of the origin at pull time could be used for optimistic concurrency purposes at push time as well.